device_backends/src/async/MuxedInterruptDistributor.cc

```diff
@@ -0,0 +1,126 @@
+// SPDX-FileCopyrightText: Deutsches Elektronen-Synchrotron DESY, MSK, ChimeraTK Project <chimeratk-support@desy.de>
+// SPDX-License-Identifier: LGPL-3.0-or-later
+#include "async/MuxedInterruptDistributor.h"
+
+#include "async/DummyMuxedInterruptDistributor.h"
+#include "async/GenericMuxedInterruptDistributor.h"
+#include "async/SubDomain.h"
+#include "async/TriggeredPollDistributor.h"
+#include <nlohmann/json.hpp>
+
+#include <tuple>
+
+namespace ChimeraTK::async {
+
+  /********************************************************************************************************************/
+
+  MuxedInterruptDistributorFactory::MuxedInterruptDistributorFactory() {
+    // we already know about the build-in handlers
+    _creatorFunctions["INTC"] = GenericMuxedInterruptDistributor::create;
+    _creatorFunctions["dummy"] = DummyMuxedInterruptDistributor::create;
+  }
+
+  /********************************************************************************************************************/
+
+  MuxedInterruptDistributorFactory& MuxedInterruptDistributorFactory::getInstance() {
+    static MuxedInterruptDistributorFactory instance;
+    return instance;
+  }
+
+  /********************************************************************************************************************/
+
+  std::pair<std::string, std::string> MuxedInterruptDistributorFactory::
+      getInterruptControllerNameAndDescriptionFromCatalogue(
+          std::vector<size_t> const& subdomainID, DeviceBackend& backend) {
+    for(auto const& metaDataEntry : backend.getMetadataCatalogue()) {
+      auto const& key = metaDataEntry.first;
+      if(key[0] == '!') {
+        auto jkey = nlohmann::json::parse(std::string({++key.begin(), key.end()})); // key without the !
+        auto interruptId = jkey.get<std::vector<size_t>>();
+
+        auto jdescriptor = nlohmann::json::parse(metaDataEntry.second);
+        auto controllerType = jdescriptor.begin().key();
+        auto controllerDescription = jdescriptor.front().dump();
+        if(interruptId == subdomainID) {
+          return {controllerType, controllerDescription};
+        }
```

Comment:
```suggestion
        if(interruptId == subdomainID) {
          auto jdescriptor = nlohmann::json::parse(metaDataEntry.second);
          auto controllerType = jdescriptor.begin().key();
          auto controllerDescription = jdescriptor.front().dump();
          return {controllerType, controllerDescription};
        }
```

But also: You are first parsing the description, then you dump it and later in the create function you parse it again, if I understand the code correctly. Shouldn't we just pass on the parsed object?

Reply:
Yes, it has to be in two steps. Here, we just get the type of controller and keep the controller-type specific description unparsed. The factory needs to know the type to know which creator function to call, but it cannot parse the description yet. This is happening in the specific creator function.

Reply:
I was just referring to the JSON parsing, of course the factory can parse the entire JSON string, it already does that. It just cannot interpret the content, but it could pass it on instead of the string.

Reply:
Ah, I could have passed on an nlohmann::json object instead of a string, right. This interface is still from the time where the string was coming directly from the catalogue, and there was no json parsing in the first step (if I remember correctly).

Reply:
On the other hand: We might want to avoid exposing the nlohmann::json object in the public API, unless we already do so, since it would add another dependency for all applications using DeviceAccess.

Reply:
That is a good point. the nlohmann is in none of the headers. I will leave it like this.